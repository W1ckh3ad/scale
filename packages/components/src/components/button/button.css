.button {
  border: 0;
  cursor: pointer;
  display: inline-flex;
  outline: none;
  position: relative;
  box-sizing: border-box;
  align-items: center;
  font-family: inherit;
  word-spacing: inherit;
  letter-spacing: inherit;
  justify-content: center;
  text-decoration: none;
  -moz-user-select: none;
  line-height: var(--type-leading-5, 1.5);
  padding-left: var(--spacing-button, 1.5rem);
  padding-right: var(--spacing-button, 1.5rem);
  font-weight: var(--font-weight-bold, 700);
  font-size: var(--font-size-3, 1rem);
  min-height: var(--size-button, 40px);
  transition: var(--transition-generic, all 0.2s ease-in-out);
}
.button:not(.button--disabled):focus {
  box-shadow: 0 0 0 var(--size-border-focus, 2px) var(--color-focus, #009de0);
}
.button.button--icon-before ::slotted(*) {
  margin-right: var(--spacing-2, 0.5rem);
}
.button.button--icon-after ::slotted(*) {
  margin-left: var(--spacing-2, 0.5rem);
}
.button:after {
  top: 0;
  left: 0;
  width: 100%;
  border: 1px solid transparent;
  height: 100%;
  content: '';
  display: block;
  position: absolute;
  box-sizing: border-box;
  pointer-events: none;
  border-radius: var(--radii-button, 0.5rem);
}
.button--size-small {
  font-size: var(--type-size-2, 0.75rem);
  line-height: var(--type-leading-4, 1.3333);
  min-height: var(--size-button-small, 32px);
}
.button--variant-primary {
  text-align: center;
  border-radius: var(--radii-button, 0.5rem);
  background: var(--color-primary, #e20074);
  color: var(--color-text-contrast, #ffffff);
}
.button--variant-primary:not(.button--disabled):hover {
  background: var(--color-primary-hover, #f90984);
}
.button--variant-primary:not(.button--disabled):active {
  background: var(--color-primary-active, #cb0068);
}
.button--disabled.button--variant-primary {
  background: var(--background-disabled, #d0d0d0);
  color: var(--background-disabled-low-contrast, #a4a4a4);
}
.button--variant-secondary {
  background: transparent;
  text-align: center;
  border-radius: var(--radii-button, 0.5rem);
  border: var(--border-size, 1px) solid currentColor;
  color: var(--color-text, #262626);
}
.button--variant-secondary:not(.button--disabled):hover {
  border-color: var(--color-primary-hover, #f90984);
  color: var(--color-primary-hover, #f90984);
}
.button--variant-secondary:not(.button--disabled):active {
  border-color: var(--color-primary-active, #cb0068);
  color: var(--color-primary-active, #cb0068);
}
.button--disabled.button--variant-secondary {
  border-color: var(--background-disabled, #d0d0d0);
  color: var(--background-disabled, #d0d0d0);
}

.button--icon-only {
  padding-left: var(--spacing-2, 0.5rem);
  padding-right: var(--spacing-2, 0.5rem);
}
.button--disabled {
  cursor: not-allowed;
}
