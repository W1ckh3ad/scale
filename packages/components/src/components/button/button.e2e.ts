import { newE2EPage } from '@stencil/core/testing';

describe('scale-button', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<scale-button>Click me!</scale-button>');
    const element = await page.find('scale-button');
    expect(element).toHaveClass('hydrated');
  });
});
