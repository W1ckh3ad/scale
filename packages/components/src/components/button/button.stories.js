import { withKnobs, boolean, text, radios } from '@storybook/addon-knobs';
import { h } from 'jsx-dom';

const ICON = 'M23.55 9.4c-.25-.7-.85-1.2-1.6-1.25l-6-.45-2.25-5.55C13.4 1.45 12.75 1 12 1s-1.4.45-1.7 1.15L8 7.75 2 8.2c-.75.05-1.35.55-1.6 1.25s0 1.45.55 1.95l4.6 3.9-1.4 5.8c-.15.7.1 1.45.7 1.9.3.25.7.35 1.05.35.35 0 .65-.1.95-.25L12 19.9l5.1 3.2c.65.4 1.4.35 2-.05.6-.45.85-1.2.7-1.9l-1.45-5.85 4.6-3.9c.6-.55.85-1.3.6-2zm-1.5.8l-5.3 4.5 1.65 6.75c.05.2-.05.3-.1.35-.05.05-.2.1-.35 0L12 18.1l-5.9 3.65c-.15.1-.3.05-.35 0-.05-.05-.15-.15-.1-.35l1.65-6.75-5.3-4.5c-.15-.1-.1-.25-.1-.35 0-.05.1-.2.25-.2l6.95-.5 2.65-6.45c.05-.15.15-.15.25-.15s.2.05.3.2l2.65 6.45 6.95.5c.2 0 .25.15.25.2 0 .1 0 .25-.1.35z';

/*
  `disabled="false"` is added everywhere to prevent a bug in the Storybook static build,
  it should be temporary and removed as soon as a fix in found.
*/

export default {
  title: 'Components/Button',
  component: 'scale-button',
  decorators: [withKnobs],
}

export const Standard = () => {
  const label = text('Label', 'Hello');

  return (
    <scale-button disabled="false">{label}</scale-button>
  );
}

export const Variants = () => {
  const label = text('Label', 'Hello');
  const variant = radios('Variant', {
    'Default': '',
    'Secondary': 'secondary'
  }, 'secondary');

  return (
    <scale-button variant={variant} disabled="false">{label}</scale-button>
  )
}

export const Disabled = () => {
  const label = text('Label', 'Hello');
  const variant = radios('Variant', {
    'Default': '',
    'Secondary': 'secondary'
  }, '');
  const disabled = boolean('Disabled', true)

  return (
    <scale-button variant={variant} disabled={disabled}>{label}</scale-button>
  )
}

export const WithIcon = () => {
  const label = text('Label', 'Hello');
  const variant = radios('Variant', {
    'Default': '',
    'Secondary': 'secondary'
  }, '');

  return (
    <scale-button icon-before={ICON} variant={variant} disabled="false">{label}</scale-button>
  )
}

export const IconOnly = () => {
  const label = text('Label', 'Hello');
  const variant = radios('Variant', {
    'Default': '',
    'Secondary': 'secondary'
  }, '');

  return (
    <scale-button icon={ICON} variant={variant} disabled="false">{label}</scale-button>
  )
}

export const Small = () => {
  const label = text('Label', 'Hello');
  const size = radios('Size', {
    'Default': '',
    'Small': 'small'
  }, 'small');

  return (
    <scale-button size={size} disabled="false">{label}</scale-button>
  )
}

export const Link = () => {
  const label = text('Label', 'Hello');
  const href = text('URL', 'https://example.com')
  const variant = radios('Variant', {
    'Default': '',
    'Secondary': 'secondary'
  }, '');

  return (
    <scale-button href={href} variant={variant} disabled="false">{label}</scale-button>
  )
}
