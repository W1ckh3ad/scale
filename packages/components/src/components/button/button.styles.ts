import { JssStyle } from 'jss';

export const styles: JssStyle = {
  button: {
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'center',
    position: 'relative',
    textDecoration: 'none',
    outline: 'none',
    cursor: 'pointer',
    userSelect: 'none',
    boxSizing: 'border-box',
    border: 0,
    paddingLeft: ({ spacing }) => spacing.x_button,
    paddingRight: ({ spacing }) => spacing.x_button,
    fontFamily: ({ type }) => type.family,
    fontWeight: ({ type }) => type.weight_bold,
    fontSize: ({ type }) => type.size_3,
    lineHeight: ({ type }) => type.leading_5,
    minHeight: ({ size }) => size.button,
    transition: ({ transition }) => transition.generic,
    // Accessibility: allows some "external" typographic styles to leak in
    all: 'inherit',

    // Focus state, the same for all variants
    '&:not($button--disabled):focus': {
      boxShadow: ({ size, color }) =>
        `0 0 0 ${size.border_focus}px ${color.focus}`,
    },

    // Spacing for inline icons
    '&$button--icon-before ::slotted(*)': {
      marginRight: ({ spacing }) => spacing[2],
    },
    '&$button--icon-after ::slotted(*)': {
      marginLeft: ({ spacing }) => spacing[2],
    },
  },

  // Accessibility: Windows High Contrast Mode transparent border
  '&:after': {
    content: '""',
    display: 'block',
    boxSizing: 'border-box',
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    border: '1px solid transparent',
    borderRadius: ({ radii }) => radii.button,
    pointerEvents: 'none',
  },

  'button--size-small': {
    fontSize: ({ type }) => type.size_2,
    lineHeight: ({ type }) => type.leading_4,
    minHeight: ({ size }) => size.button_small,
  },

  'button--variant-primary': {
    borderRadius: ({ radii }) => radii.button,
    background: ({ color }) => color.primary,
    color: ({ color }) => color.text_contrast,
    textAlign: 'center',

    '&:not($button--disabled):hover': {
      background: ({ color }) => color.primary_hover,
    },
    '&:not($button--disabled):active': {
      background: ({ color }) => color.primary_active,
    },
    '$button--disabled&': {
      background: ({ background }) => background.disabled,
      color: ({ color }) => color.disabled,
    },
  },

  'button--variant-secondary': {
    borderRadius: ({ radii }) => radii.button,
    border: ({ size }) => `${size.border}px solid currentColor`,
    background: 'transparent',
    color: ({ color }) => color.text,
    textAlign: 'center',

    '&:not($button--disabled):hover': {
      borderColor: ({ color }) => color.primary_hover,
      color: ({ color }) => color.primary_hover,
    },
    '&:not($button--disabled):active': {
      borderColor: ({ color }) => color.primary_active,
      color: ({ color }) => color.primary_active,
    },
    '$button--disabled&': {
      borderColor: ({ color }) => color.disabled,
      color: ({ color }) => color.disabled,
    },
  },

  'button--icon-before': {},
  'button--icon-after': {},

  'button--icon-only': {
    paddingLeft: ({ spacing }) => spacing[2],
    paddingRight: ({ spacing }) => spacing[2],
  },

  'button--disabled': {
    cursor: 'not-allowed',
  },
};
