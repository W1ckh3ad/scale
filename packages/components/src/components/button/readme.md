# scale-button



<!-- Auto Generated Below -->


## Properties

| Property       | Attribute       | Description                                                    | Type                              | Default     |
| -------------- | --------------- | -------------------------------------------------------------- | --------------------------------- | ----------- |
| `customClass`  | `custom-class`  | (optional) Custom class                                        | `string`                          | `''`        |
| `disabled`     | `disabled`      | (optional) If `true`, the button is disabled                   | `boolean`                         | `false`     |
| `href`         | `href`          | (optional) When present, an <a> tag will be used               | `string`                          | `undefined` |
| `iconOnly`     | `icon-only`     | (optional) Set to `true` when the button contains only an icon | `boolean`                         | `false`     |
| `iconPosition` | `icon-position` | (optional) Icon position related to the label                  | `"after" \| "before"`             | `'before'`  |
| `size`         | `size`          | (optional) The size of the button                              | `"large" \| "small"`              | `'large'`   |
| `styles`       | `styles`        | (optional) Injected jss styles                                 | `any`                             | `undefined` |
| `target`       | `target`        | (optional) The target attribute for the <a> tag                | `string`                          | `'_self'`   |
| `type`         | `type`          | (optional) Button type                                         | `"button" \| "reset" \| "submit"` | `undefined` |
| `variant`      | `variant`       | (optional) Button variant                                      | `string`                          | `'primary'` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
