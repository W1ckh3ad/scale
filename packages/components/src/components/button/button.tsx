import { Component, Prop, h, Host, Element } from '@stencil/core';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';
import { styles } from './button.styles';
import { CssInJs } from '../../utils/css-in-js';
import { hasShadowDom } from '../../utils/utils';
import { StyleSheet } from 'jss';
import Base from '../../utils/base-interface';

/*
  TODO
  - [ ] update storybook (remove link disabled?)
*/

@Component({
  tag: 'scale-button',
  shadow: true,
})
export class Button implements Base {
  @Element() hostElement: HTMLElement;

  /** (optional) Custom class */
  @Prop() customClass?: string = '';
  /** (optional) The size of the button */
  @Prop() size?: 'small' | 'large' = 'large';
  /** (optional) Button variant */
  @Prop() variant?: string = 'primary';
  /** (optional) If `true`, the button is disabled */
  @Prop() disabled?: boolean = false;
  /** (optional) Button type */
  @Prop() type?: 'reset' | 'submit' | 'button';
  /** (optional) Set to `true` when the button contains only an icon */
  @Prop() iconOnly?: boolean = false;
  /** (optional) Icon position related to the label */
  @Prop({ reflect: true }) iconPosition: 'before' | 'after' = 'before';
  /** (optional) aria-label attribute needed for icon-only buttons */
  @Prop() ariaLabel: string;
  /** (optional) When present, an <a> tag will be used */
  @Prop() href?: string;
  /** (optional) The target attribute for the <a> tag */
  @Prop() target?: string = '_self';

  /** (optional) Injected jss styles */
  @Prop() styles?: any;
  /** decorator Jss stylesheet */
  @CssInJs('Button', styles) stylesheet: StyleSheet;

  /**
   * Hack to make the button behave has expected when inside forms.
   * @see https://github.com/ionic-team/ionic-framework/blob/master/core/src/components/button/button.tsx#L155-L175
   */
  handleClick = (ev: Event) => {
    if (hasShadowDom(this.hostElement) && this.disabled === false) {
      const form = this.hostElement.closest('form');
      if (form) {
        ev.preventDefault();

        const fakeButton = document.createElement('button');
        if (this.type) {
          fakeButton.type = this.type;
        }
        fakeButton.style.display = 'none';
        form.appendChild(fakeButton);
        fakeButton.click();
        fakeButton.remove();
      }
    }
  };

  componentWillUpdate() {}
  componentDidUnload() {}

  connectedCallback() {
    this.setIconPositionProp();
  }

  /**
   * Detect whether the last node is an element (not text).
   * If so, it's probably an icon, so we set `iconPosition` to `after`.
   */
  setIconPositionProp() {
    const nodes = Array.from(this.hostElement.childNodes);
    if (nodes.length < 2) {
      return;
    }
    const lastNode = nodes[nodes.length - 1];
    if (lastNode != null && lastNode.nodeType === 1) {
      this.iconPosition = 'after';
    }
  }

  render() {
    return (
      <Host>
        <style>{this.stylesheet.toString()}</style>
        {this.href ? (
          <a
            class={this.getCssClassMap()}
            href={this.href}
            target={this.target}
            rel={this.target === '_blank' ? 'noopener noreferrer' : undefined}
            aria-label={this.ariaLabel}
          >
            <slot />
          </a>
        ) : (
          <button
            class={this.getCssClassMap()}
            onClick={this.handleClick}
            disabled={this.disabled}
            type={this.type}
            aria-label={this.ariaLabel}
          >
            <slot />
          </button>
        )}
      </Host>
    );
  }

  getCssClassMap(): CssClassMap {
    const { classes } = this.stylesheet;
    return classNames(
      classes.button,
      this.customClass && classes[this.customClass],
      this.size && classes[`button--size-${this.size}`],
      this.variant && classes[`button--variant-${this.variant}`],
      this.iconOnly && classes[`button--icon-only`],
      !this.iconOnly &&
        this.iconPosition &&
        classes[`button--icon-${this.iconPosition}`],
      this.disabled && !this.href && classes[`button--disabled`]
    );
  }
}
