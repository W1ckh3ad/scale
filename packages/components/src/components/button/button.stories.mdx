import { Meta, Story, Description, Preview } from '@storybook/addon-docs/blocks';
import * as stories from './button.stories.js';
import readme from './readme.md'

<Meta title="Components/Button" component="scale-button"/>

# Scale Button

<Preview>
  <Story name="Standard">
    {stories.Standard()}
  </Story>
</Preview>

<Preview>
  <Story name="Variants">
    {stories.Variants()}
  </Story>
</Preview>

<Preview>
  <Story name="Disabled">
    {stories.Disabled()}
  </Story>
</Preview>

<Preview>
  <Story name="With Icon">
    {stories.WithIcon()}
  </Story>
</Preview>

<Preview>
  <Story name="Icon Only">
    {stories.IconOnly()}
  </Story>
</Preview>

<Preview>
  <Story name="Small">
    {stories.Small()}
  </Story>
</Preview>

<Preview>
  <Story name="Link">
    {stories.Link()}
  </Story>
</Preview>

<Description>
  {readme}
</Description>
