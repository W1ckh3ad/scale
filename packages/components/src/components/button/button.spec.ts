import { newSpecPage } from '@stencil/core/testing';
import { Button } from './button';
import { styles } from './button.styles';
import jss from 'jss';

describe('Button', () => {
  let element;
  let stylesheet;
  beforeEach(async () => {
    element = new Button();
    stylesheet = element.stylesheet = jss.createStyleSheet(styles as any);
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<scale-button>Label</scale-button>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should disable the button', () => {
    expect(element.disabled).toBe(false);
    element.disable();
    expect(element.disabled).toBe(true);
  });

  it('should enable the button', () => {
    element.disable();
    expect(element.disabled).toBe(true);
    element.enable();
    expect(element.disabled).toBe(false);
  });

  it('should handle custom css class', () => {});

  it('should handle css classes', () => {
    element.customClass = 'custom';
    expect(element.getCssClassMap()).toContain('custom');

    element.size = 'small';
    stylesheet.addRule('button--size-small', {});
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['button--size-small']
    );

    element.variant = 'primary';
    stylesheet.addRule('button--variant-primary', {});
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['button--variant-primary']
    );

    element.disabled = true;
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['button--disabled']
    );
  });
});
