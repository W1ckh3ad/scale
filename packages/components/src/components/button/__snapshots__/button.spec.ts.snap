// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot 1`] = `
<scale-button>
  <mock:shadow-root>
    <style>
      .button-0-0-8 {
  border: none;
  cursor: pointer;
  height: unset;
  display: inline-flex;
  outline: none;
  position: relative;
  box-shadow: none;
  box-sizing: border-box;
  min-height: 40px;
  text-align: center;
  align-items: center;
  flex-shrink: 0;
  padding-top: 8px;
  user-select: none;
  white-space: nowrap;
  padding-left: 24px;
  border-radius: 8px;
  padding-right: 24px;
  letter-spacing: 0.02857em;
  padding-bottom: 8px;
  text-transform: none;
  vertical-align: middle;
  justify-content: center;
  text-decoration: none;
  color: #000;
  --icon-color: #000;
  background: #E0E0E0;
  font-family: TeleNeoWeb;
  font-weight: 700;
  font-size: 1rem;
  line-height: 1.5;
  transition: all 0.2s cubic-bezier(0.4, 0, 0.2, 1);
}
.button-0-0-8:before {
  width: 0;
  height: 100%;
  content: "";
  display: block;
}
.button-0-0-8:hover {
  box-shadow: none;
  text-decoration: none;
  transition: all 0.3s cubic-bezier(0.4, 0, 0.2, 1);
  border-color: #F90984;
  background: #F90984;
  color: #FFF;
  --icon-color: #FFF;
}
.button-0-0-8, .button-0-0-8:focus {
  border-color: #E20074;
  background: #E20074;
  color: #FFF;
  --icon-color: #FFF;
}
.button-0-0-8:focus {
  box-shadow: 0 0 0 2px #009DE0;
}
.button-0-0-8:active, .button-0-0-8.selected {
  border-color: #CB0068;
  background: #CB0068;
  color: #FFF;
  --icon-color: #FFF;
}
.button--disabled-0-0-9 {}
.button--disabled-0-0-9, .button--disabled-0-0-9:hover {
  cursor: not-allowed;
  opacity: 1;
  background: #D0D0D0;
  color: #7C7C7C;
  --icon-color: #7C7C7C;
}
.button--disabled-0-0-9.button--variant-secondary-0-0-13, .button--disabled-0-0-9.button--variant-secondary-0-0-13:hover {
  background: none;
  border: 1px solid #D0D0D0;
  color: #7C7C7C;
  --icon-color: #7C7C7C;
}
.button__before-0-0-10 {
  display: flex;
  align-items: center;
  margin-right: .5rem;
}
.button__after-0-0-11 {
  display: flex;
  align-items: center;
  margin-left: .5rem;
}
.button--icon-only-0-0-12 {
  padding: 8px;
}
.button--variant-secondary-0-0-13 {
  padding-top: 7px;
  padding-bottom: 7px;
}
.button--variant-secondary-0-0-13, .button--variant-secondary-0-0-13:focus, .button--variant-secondary-0-0-13.focus {
  background: none;
  border: 1px solid #262626;
  color: #262626;
  --icon-color: #262626;
}
.button--variant-secondary-0-0-13:hover, .button--variant-secondary-0-0-13.hover {
  background: none;
  border-color: #F90984;
  color: #F90984;
  --icon-color: #F90984;
}
.button--variant-secondary-0-0-13:active, .button--variant-secondary-0-0-13.active, .button--variant-secondary-0-0-13.selected {
  background: none;
  border-color: #CB0068;
  color: #CB0068;
  --icon-color: #CB0068;
}
.button--size-small-0-0-14 {
  min-height: 32px;
}
.button--size-small-0-0-14.button--icon-only-0-0-12 {
  min-width: 32px;
  min-height: 32px;
}
    </style>
    <button class="button-0-0-8" tabindex="0">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Label
</scale-button>
`;
